scenario/state.py

```diff
@@ -120,6 +122,26 @@ class MetadataNotFoundError(RuntimeError):
     """Raised when Scenario can't find a metadata.yaml file in the provided charm root."""
 
 
+# This can be replaced with the KW_ONLY dataclasses functionality in Python 3.10+.
+class _MaxPositionalArgs:
+    """Raises TypeError when instantiating objects if arguments are not passed as keywords.
+
+    Looks for a `_max_positional_args` class attribute, which should be an int
+    indicating the maximum number of positional arguments that can be passed to
+    `__init__` (excluding `self`). If not present, no limit is applied.
+    """
+
+    _max_positional_args = 0
+
+    def __new__(cls, *args, **_):
+        if len(args) > getattr(cls, "_max_positional_args", float("inf")):
```

Comment:
in what circumstances will the getattr fall back to the default? given this class has the attribute, all subclasses will inherit it. 
Unless someone goes exceptionally out of their way and `delattr`s it, it should always be there, no?

Reply:
Yes, fair point. I think maybe I didn't have it in the parent class originally and this was leftover from that. I'll clean it up, thanks!